### PR TITLE
govc: Add support for CNS volume snapshots

### PR DESCRIPTION
**Pull request by grandeit**
## Description

Adds support for interacting with snapshots of CNS volumes by implementing the following commands:
- `govc volume.snapshot.create ID DESC`
Creates a snapshot of the CNS volume with volume id `ID` with the description `DESC`.
- `govc volume.snapshot.ls ID`
Lists all snaphots of the CNS volume with volume id `ID`
- `govc volume.snapshot.rm SNAP_ID VOL_ID`
Removes the snapshot with snapshot id `SNAP_ID` from the CNS volume with the volume id `VOL_ID`

The commands `volume.snapshot.ls` and `volume.snapshot.rm` support batch operations that make it possible to list all snapshots of multiple volumes or delete multiple snapshots at once. For example it's possible to delete all snapshots of all volumes using `govc volume.snapshot.rm $(govc volume.snapshot.ls -i $(govc volume.ls -i))`.

All commands return the results of the operation, i.e. listing the delted or created snapshots.

On error, the commands `volume.snapshot.ls` and `volume.snapshot.create` return immediatly while the command `volume.snapshot.rm` returns after all the input has been processed. This is done to be able to provide feedback in case only a few delete operation from a larger list fail.

Only a very basic BATS test case has been added since the standard vsim environment does not contain a CNS volume which could have been used for snapshot operations.

## Type of change

Please mark options that are relevant:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to
  not work as expected)
- [ ] This change requires a documentation update
- [ ] Build related change

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide
instructions so we can reproduce. If applicable, please also list any relevant
details for your test configuration.

Tested on a live vSphere cluster running vCenter v7.0.3

- [x] List all snapshots from all CNS volumes
```
$ govc volume.snapshot.ls -l $(govc volume.ls -i)
4c35232b-692a-4008-9e14-99dfab529bed  snapshot-535503cf-547c-435f-9222-39d8910f0576-d6d6ff5f-b2eb-4446-9ce2-980d65d6cfc7  d6d6ff5f-b2eb-4446-9ce2-980d65d6cfc7  Mar 15 13:45:52
ae385bb0-81b9-4b91-82dd-10d20afc9961  snapshot-8e21ff38-380a-4b7a-8c59-3e14387a9768-d6d6ff5f-b2eb-4446-9ce2-980d65d6cfc7  d6d6ff5f-b2eb-4446-9ce2-980d65d6cfc7  Mar 15 14:09:05
4ad63cae-b0a5-4eca-8df9-4d6ed2913c50  snapshot-a98aa867-7ef5-45a6-ba14-b28a145ab867-d6d6ff5f-b2eb-4446-9ce2-980d65d6cfc7  d6d6ff5f-b2eb-4446-9ce2-980d65d6cfc7  Mar 15 14:13:16
1d8e401b-c1b7-4771-802a-050e892166db  snapshot-f00c21de-1068-4f73-9d2d-4d7846d203d5-bf10a2c8-23e5-4bcb-983f-2358d9f77151  bf10a2c8-23e5-4bcb-983f-2358d9f77151  Mar 20 06:04:17
b6c248d5-c8b5-4059-a732-83b9715381de  snapshot-5287f7f6-11ff-4c93-9ade-4405ca18b3c4-31833215-4b10-4e2d-8cd4-3a9fd2c4f8af  31833215-4b10-4e2d-8cd4-3a9fd2c4f8af  Mar 27 15:43:46
66aa3e6e-467b-41c8-b7f5-6dfa8de95058  snapshot-60fff95b-ee35-4b1a-8917-3cec9b362486-31833215-4b10-4e2d-8cd4-3a9fd2c4f8af  31833215-4b10-4e2d-8cd4-3a9fd2c4f8af  Mar 28 15:32:18
3d8636d3-b7eb-4e8a-8428-4c0134473e27  snapshot-46213458-8ec2-483b-b7a1-651a50a13af1-8568053e-4d6b-4684-af5f-8505d6f98b0e  8568053e-4d6b-4684-af5f-8505d6f98b0e  Mar 15 13:46:02
7afc8c06-9c48-4c73-911b-f38333779d39  snapshot-2f614f9a-89fc-475a-8f56-2d8d4f768f8d-8568053e-4d6b-4684-af5f-8505d6f98b0e  8568053e-4d6b-4684-af5f-8505d6f98b0e  Mar 15 14:09:15
66bbc2e1-aeee-4c33-95a3-b96a91610daa  snapshot-9b6d68bc-f5e8-4eee-b97c-fa8fc007c6b6-8568053e-4d6b-4684-af5f-8505d6f98b0e  8568053e-4d6b-4684-af5f-8505d6f98b0e  Mar 15 14:13:26
35fea4f4-a074-4455-91a0-d3db0308f8c2  snapshot-b48c5cf7-399d-411c-83e7-333c67c95afd-1914fee1-9030-44c2-8444-a5a4928100e5  1914fee1-9030-44c2-8444-a5a4928100e5  Mar 27 15:45:53
000f81a8-6f01-4996-92e9-b916335b82f6  snapshot-9fa75bde-42c2-4b79-8c45-7bf2ebb59b46-1b929050-75a0-48a3-9de0-ecfdc0418440  1b929050-75a0-48a3-9de0-ecfdc0418440  Mar 27 15:48:19
8d0e27f1-82fe-4bd1-a479-38e31aee9c02  snapshot-1c107c91-b6b9-43f5-ad89-8375c272a814-ad44e8b2-a8f5-4940-abca-82f6c8edbfad  ad44e8b2-a8f5-4940-abca-82f6c8edbfad  Mar 27 15:52:50
```
- [x] List all snapshots from a specific CNS volume
```
$ govc volume.snapshot.ls -l d6d6ff5f-b2eb-4446-9ce2-980d65d6cfc7
4c35232b-692a-4008-9e14-99dfab529bed  snapshot-535503cf-547c-435f-9222-39d8910f0576-d6d6ff5f-b2eb-4446-9ce2-980d65d6cfc7  d6d6ff5f-b2eb-4446-9ce2-980d65d6cfc7  Mar 15 13:45:52
ae385bb0-81b9-4b91-82dd-10d20afc9961  snapshot-8e21ff38-380a-4b7a-8c59-3e14387a9768-d6d6ff5f-b2eb-4446-9ce2-980d65d6cfc7  d6d6ff5f-b2eb-4446-9ce2-980d65d6cfc7  Mar 15 14:09:05
4ad63cae-b0a5-4eca-8df9-4d6ed2913c50  snapshot-a98aa867-7ef5-45a6-ba14-b28a145ab867-d6d6ff5f-b2eb-4446-9ce2-980d65d6cfc7  d6d6ff5f-b2eb-4446-9ce2-980d65d6cfc7  Mar 15 14:13:16
```
- [x] Create a snapshot on a CNS volume
```
$ govc volume.snapshot.ls -l d3c769bd-bb70-474b-bc31-fbe005a013ac
$ govc volume.snapshot.create d3c769bd-bb70-474b-bc31-fbe005a013ac my-snapshot
d63b1690-f049-4d28-a8be-35f0ce5fb058  my-snapshot  d3c769bd-bb70-474b-bc31-fbe005a013ac  Apr  1 21:02:19
$ govc volume.snapshot.ls -l d3c769bd-bb70-474b-bc31-fbe005a013ac
d63b1690-f049-4d28-a8be-35f0ce5fb058  my-snapshot  d3c769bd-bb70-474b-bc31-fbe005a013ac  Apr  1 21:02:19
```
- [x] Remove a snapshot from a CNS volume
```
$ govc volume.snapshot.ls -l d3c769bd-bb70-474b-bc31-fbe005a013ac
d63b1690-f049-4d28-a8be-35f0ce5fb058  my-snapshot  d3c769bd-bb70-474b-bc31-fbe005a013ac  Apr  1 21:02:19
$ govc volume.snapshot.rm d63b1690-f049-4d28-a8be-35f0ce5fb058 d3c769bd-bb70-474b-bc31-fbe005a013ac
d63b1690-f049-4d28-a8be-35f0ce5fb058  d3c769bd-bb70-474b-bc31-fbe005a013ac
$ govc volume.snapshot.ls -l d3c769bd-bb70-474b-bc31-fbe005a013ac
$ 
```
- [x] Remove a list of snapshots from a CNS volume
```
$ govc volume.snapshot.ls -l d3c769bd-bb70-474b-bc31-fbe005a013ac
$ govc volume.snapshot.create d3c769bd-bb70-474b-bc31-fbe005a013ac my-snapshot-001
1ad5fd95-f804-4065-baae-3a59aefc0d14  my-snapshot-001  d3c769bd-bb70-474b-bc31-fbe005a013ac  Apr  1 21:07:00
$ govc volume.snapshot.create d3c769bd-bb70-474b-bc31-fbe005a013ac my-snapshot-002
792dbf5f-8d9b-4d6f-b8b3-80d75f31548b  my-snapshot-002  d3c769bd-bb70-474b-bc31-fbe005a013ac  Apr  1 21:07:08
$ govc volume.snapshot.create d3c769bd-bb70-474b-bc31-fbe005a013ac my-snapshot-003
3e002e3c-3e16-4c01-88d5-8a0ab9601838  my-snapshot-003  d3c769bd-bb70-474b-bc31-fbe005a013ac  Apr  1 21:07:16
$ govc volume.snapshot.create d3c769bd-bb70-474b-bc31-fbe005a013ac my-snapshot-004
47a2d468-cc99-43d1-bf14-0b252f54b3d9  my-snapshot-004  d3c769bd-bb70-474b-bc31-fbe005a013ac  Apr  1 21:07:24
$ govc volume.snapshot.ls -l d3c769bd-bb70-474b-bc31-fbe005a013ac
1ad5fd95-f804-4065-baae-3a59aefc0d14  my-snapshot-001  d3c769bd-bb70-474b-bc31-fbe005a013ac  Apr  1 21:07:00
792dbf5f-8d9b-4d6f-b8b3-80d75f31548b  my-snapshot-002  d3c769bd-bb70-474b-bc31-fbe005a013ac  Apr  1 21:07:08
3e002e3c-3e16-4c01-88d5-8a0ab9601838  my-snapshot-003  d3c769bd-bb70-474b-bc31-fbe005a013ac  Apr  1 21:07:16
47a2d468-cc99-43d1-bf14-0b252f54b3d9  my-snapshot-004  d3c769bd-bb70-474b-bc31-fbe005a013ac  Apr  1 21:07:24
$ govc volume.snapshot.rm $(govc volume.snapshot.ls -i d3c769bd-bb70-474b-bc31-fbe005a013ac)
1ad5fd95-f804-4065-baae-3a59aefc0d14  d3c769bd-bb70-474b-bc31-fbe005a013ac
792dbf5f-8d9b-4d6f-b8b3-80d75f31548b  d3c769bd-bb70-474b-bc31-fbe005a013ac
3e002e3c-3e16-4c01-88d5-8a0ab9601838  d3c769bd-bb70-474b-bc31-fbe005a013ac
47a2d468-cc99-43d1-bf14-0b252f54b3d9  d3c769bd-bb70-474b-bc31-fbe005a013ac
$ govc volume.snapshot.ls -l d3c769bd-bb70-474b-bc31-fbe005a013ac
$ 
```
## Checklist:

- [x] My code follows the `CONTRIBUTION`
  [guidelines](https://github.com/vmware/govmomi/blob/main/CONTRIBUTING.md) of
  this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged

### CODE REVIEW

**Comment by grandeit** [comment]
I used the `contributors.sh`, `license.sh` and `usage.sh` scripts to update the relevant files.

**Comment by grandeit** [comment]
Also fixed some minor styling issues.

**Review by dougm — APPROVED** [review]
Thanks @grandeit , lgtm. Tagging @divyenpatel and @SandeepPissay for review

File: govc/volume/snapshot/create.go [hunk 1]
```diff
@@ -0,0 +1,129 @@
+/*
+Copyright (c) 2023 VMware, Inc. All Rights Reserved.
```
Comment (dougm) [inline govc/volume/snapshot/create.go:2]: linter expects a range even for single year, `2023-2023`. I've an update to license.sh for this in a pending PR #3083 3083

Comment (grandeit) [reply govc/volume/snapshot/create.go]: Fixed.

**Review by dougm — APPROVED** [review]
Thanks again @grandeit

**Pull request merged**